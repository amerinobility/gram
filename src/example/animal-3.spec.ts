import { GraphQLBoolean, GraphQLInt, GraphQLString, printSchema } from 'graphql'
import { createSchemaBuilder, ListType, NodeType, Service } from '..'

interface Animal extends NodeType {
  name: string
  mother: Animal
  father: Animal
  children: Animal[]
}

const Animals: Service<Animal> = {
  findOne: async () => null,
  findMany: async () => ({
    page: null,
    nodes: [],
  }),
}

const AnimalService: Service<Animal> = {
  findOne: async ({ order, where }, context) =>
    Animals.findOne({ order, where }, context) as Promise<Animal>,
  findMany: async ({ order, page, where }, context) =>
    Animals.findMany({ order, page, where }, context) as Promise<
      ListType<Animal>
    >,
}

describe('testing the example 3', () => {
  it('should build the example code', () => {
    const builder = createSchemaBuilder()
    const animal = builder.model('Animal', AnimalService)
    // animal name like 'Fluffy', 'Rex'
    // field is required
    animal.attr('name', 'String!')
    // parents
    animal.attr('mother', 'Animal')
    animal.attr('father', 'Animal')
    // children
    animal.attr('children', '[Animal!]!')

    const schema = builder.build()

    expect(printSchema(schema)).toMatchSnapshot()
  })
})
