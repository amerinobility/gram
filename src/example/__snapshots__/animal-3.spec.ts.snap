// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the example 3 should build the example code 1`] = `
"type Animal implements Node {
  name: String!
  mother: Animal
  father: Animal
  children: [Animal!]!
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input AnimalFilter {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  children_contains: AnimalWhere
  children_not_contains: AnimalWhere
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input AnimalPage {
  limit: Int
  offset: Int
}

type Animals implements List {
  page: Page
  nodes: [Animal!]!
}

enum AnimalSortOrder {
  name_ASC
  name_DESC
  mother_ASC
  mother_DESC
  father_ASC
  father_DESC
  children_ASC
  children_DESC
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
}

input AnimalWhere {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  children_contains: AnimalWhere
  children_not_contains: AnimalWhere
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input CreateAnimalData {
  name: String!
  mother: AnimalWhere
  father: AnimalWhere
  children: AnimalWhere!
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

scalar DateTime

interface List {
  page: Page
  nodes: [Node!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getAnimal(where: AnimalWhere!, order: AnimalSortOrder): Animal
  getAnimals(order: AnimalSortOrder, page: AnimalPage, where: AnimalWhere!): Animals
}

input UpdateAnimalData {
  name: String
  mother: AnimalWhere
  father: AnimalWhere
  children: AnimalWhere
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}
"
`;
